### DISCUSSION TITLE
Using user response in subsequent survey question

### DISCUSSION THREAD

**Discussion by Kedersha** [jspsych/jsPsych#2658] [Q&A] [answered]
Hello! I'm trying to set up a survey where first, participants provide a short list of languages they speak, and then second, they rank that list in order of proficiency. I can pull the relevant responses in the console, but I can't use them in the actual study code because it throws this error: `Uncaught TypeError: jsPsych.data.getLastTrialData().trials[0] is undefined`

What's the correct syntax for using user input as options in a subsequent question?

Here's the relevant code; I've tried all of the commented-out bits as members in the "options" array, to no avail:

```
	//Languages
    var lang_all = {
      type: jsPsychSurvey,
	  stimulus: '',
      pages: [
        [
		  {type: 'html', prompt: header + "<p>Please list the languages you use in your life; if you know more than 5 languages, please list the 5 you use the most.</p>"},
          {type: 'text', prompt: 'Language 1:', name: 'lang1', required: true},
		  {type: 'text', prompt: 'Language 2:', name: 'lang2'},
		  {type: 'text', prompt: 'Language 3:', name: 'lang3'},
		  {type: 'text', prompt: 'Language 4:', name: 'lang4'},
		  {type: 'text', prompt: 'Language 5:', name: 'lang5'}
		]
    ]
    };
	
	var lang_rank = {
		type: jsPsychSurvey,
		stimulus: '',
		pages: [
			[
          {type: 'ranking', prompt: 'Proficiency with languages', options: ["English",jsPsych.data.getLastTrialData().trials[0].response.lang1]}
		  //jsPsych.data.get().last(1).values()[0].response.lang1
		  //jsPsych.data.getLastTrialData().trials[0].response.lang1 
		  //JSON.parse(jsPsych.data.getLastTrialData().select('responses').values).lang1
        ]
	]
	};
```

**Comment by sophief10** [comment, marked as answer]
Hi! Here is my understanding of the problem:

jsPsych will evaluate all parameter values as soon as the experiment loads, even if those parameters are for trials that occur at the end of the experiment. It assumes that the programmer (you) has set the content of each experimental trial to be the same for every study participant, so the parameters of every trial become permanent as soon as the experiment starts and will not respond to data that the participant produces during the experiment. 

You can get around this obstacle by setting the parameter values to be functions ([dynamic parameters](https://www.jspsych.org/7.2/overview/dynamic-parameters/)). jsPsych only evaluates functions when it comes time to use their outputs, so if a parameter is equal to a function, jsPsych will evaluate the parameter only just before the trial starts, and will use the function output as the parameter value.

Your code works fine on my computer when I change it to this:

```
var lang_rank = {
  pages: function() {
    return [
      [
        {type: 'ranking', prompt: 'Proficiency with languages', options: ['English', jsPsych.data.getLastTrialData().trials[0].response.lang1]}
      ]
    ]
  }
}
```

**Reply by CarineA** [reply]
Hi @puzzleGuzzle I'm having a similar issue where it says that "TypeError: Cannot read properties of undefined (reading 'response')"

I have a really long survey, so I'm guessing it might be an issue that there are many pages? Any thoughts on this? 

`
var LSBQ = {
  type: jsPsychSurvey,
  pages: function(){
    return [
    [
//Other pages here including the question I'm pulling from
  {
          type: "text",
          prompt: "Language 1:",
          name: "L1",
          required: false,
        },
    ],
      [
        {
        type:'html',
        prompt: '<h3 style="text-align: center; color: black;">'+jsPsych.data.getLastTrialData().values()[0].response.L1+'</h3>'
      },
]
]`

**Reply by CarineA** [reply]
I figured it out! It just didn't like being in the same variable

**Comment by Kedersha** [comment]
Thanks for your clear and concise explanation and demonstration! Things are working well for me now. 👍
